lightning/src/ln/channelmanager.rs

```diff
@@ -2938,8 +2942,34 @@ macro_rules! handle_error {
 ///
 /// Note that this step can be skipped if the channel was never opened (through the creation of a
 /// [`ChannelMonitor`]/channel funding transaction) to begin with.
-macro_rules! update_maps_on_chan_removal {
-	($self: expr, $peer_state: expr, $channel_context: expr) => {{
+macro_rules! locked_close_channel {
+	($self: ident, $peer_state: expr, $channel_context: expr, $shutdown_res_mut: expr) => {{
+		if let Some((counterparty_node_id, funding_txo, channel_id, update)) = $shutdown_res_mut.monitor_update.take() {
+			if $self.background_events_processed_since_startup.load(Ordering::Acquire) {
+				handle_new_monitor_update!($self, funding_txo, update, $peer_state,
+					$channel_context, REMAIN_LOCKED_UPDATE_ACTIONS_PROCESSED_LATER);
+			} else {
+				// We want to track the in-flight update both in `in_flight_monitor_updates` and in
+				// `pending_background_events` to avoid a race condition during
+				// `pending_background_events` processing where we complete one
+				// `ChannelMonitorUpdate` (but there are more pending) but we conclude that all
+				// pending `ChannelMonitorUpdate`s have completed and its safe to run
+				// post-completion actions. We could work around that with some effort, but its
+				// simpler to just track updates twice.
+				let in_flight_updates = $peer_state.in_flight_monitor_updates.entry(funding_txo)
+					.or_insert_with(Vec::new);
+				if !in_flight_updates.contains(&update) {
+					in_flight_updates.push(update.clone());
+				}
```

Comment:
Thinking more on this race condition to make sure I'm following. Is one possible consequence of running the post-completion actions early that we may free an upstream channel before a preimage is persisted on the downstream channel? If there's a money-losing condition like that, it does seem worth testing, currently commenting out pushing the `in_flight_update` lines passes tests. Not gonna hold up the PR on it though.